### PR TITLE
Fix/miner fixes

### LINKED ISSUE #409: Game crash on farm land block state change?
### Minecolonies version

version 0.6.2190

### Logged Error

get property PropertyEnum{name=variant, clazz=class net.minecraft.block.BlockDirt$DirtType, values=[dirt, coarse_dirt, podzol]} as it does not exist in BlockStateContainer{block=minecraft:farmland, properties=[moisture]}

### Log Files
[fml-server-3.zip](https://github.com/Minecolonies/minecolonies/files/680217/fml-server-3.zip)
[McMyAdmin-20170101-162037.zip](https://github.com/Minecolonies/minecolonies/files/680218/McMyAdmin-20170101-162037.zip)

Also noticed that a lot of the labels for items and achievements are not in a user friendly format on this build.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #409, #410 

# Changes proposed in this pull request:
- Miner should now navigate way more smoothly.
- Fix some problems with colonist spawning.
- Fix some more minor problems.
Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 1]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
             //If he is on the same Y level as his target and both underground, don't use a proxy. Just don't.
             else if (targetY <= levelDepth)
             {
+                double closestNode = Double.MAX_VALUE;
+                Node lastNode = null;
+                for(Map.Entry<Point2D, Node> node : level.getNodes().entrySet())
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR Local variable 'node' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 2]
```diff
@@ -92,8 +106,14 @@ private boolean takeTheDirectPath(@NotNull BlockPos target, int range, boolean o
...
     {
-        final double distanceToPath = worker.getColonyJob() instanceof JobBuilder
-                ? BlockPosUtil.getDistanceSquared2D(worker.getPosition(), target) : BlockPosUtil.getDistanceSquared(worker.getPosition(), target);
+        if(!target.equals(this.target))
+        {
+            reset();
+            this.target = target;
+        }
+
+        final double distanceToPath = worker.getColonyJob() instanceof JobBuilder ?
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR '?' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 3]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
             //If he is on the same Y level as his target and both underground, don't use a proxy. Just don't.
             else if (targetY <= levelDepth)
             {
+                double closestNode = Double.MAX_VALUE;
+                Node lastNode = null;
+                for(Map.Entry<Point2D, Node> node : level.getNodes().entrySet())
+                {
+                    double distanceToNode = node.getKey().distance(worker.getPosition().getX(), worker.getPosition().getZ());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR Local variable 'distanceToNode' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 4]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
...
+                        currentNode = level.getNode(currentNode.getParent());
+                    }
+
+                    for (int i = nodesToTarget.size() - 1; i >= 0; i--)
+                    {
+                        proxyList.add(nodesToTarget.get(i));
+                    }
+                }
+
+                if(proxyList.size() != 0)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1155)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 5]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
...
+                    {
+                        proxyList.add(new BlockPos(currentNode.getX(), levelDepth, currentNode.getZ()));
+                        currentNode = level.getNode(currentNode.getParent());
+                    }
+                }
+
+                if(level.getRandomNode().getParent() != null)
+                {
+                    final List<BlockPos> nodesToTarget = new ArrayList<>();
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(level.getRandomNode().getParent());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:312]: MAJOR Unnecessary use of fully qualified name 'com.minecolonies.coremod.entity.ai.citizen.miner.Node' due to existing import 'com.minecolonies.coremod.entity.ai.citizen.miner.*' [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 6]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
...
+                    if(distanceToNode < closestNode)
+                    {
+                        lastNode = node.getValue();
+                        closestNode = distanceToNode;
+                    }
+                }
+
+                if(lastNode != null && lastNode.getParent() != null)
+                {
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(lastNode.getParent());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:301]: MAJOR Unnecessary use of fully qualified name 'com.minecolonies.coremod.entity.ai.citizen.miner.Node' due to existing import 'com.minecolonies.coremod.entity.ai.citizen.miner.*' [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* MAJOR 5 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/d8c6abc32501a4e349e22cff13fd4b3348349122/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/d8c6abc32501a4e349e22cff13fd4b3348349122/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* MAJOR 5 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/e4d0bef94ab9f5d30a72f76c71e8682d32ce198c/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/e4d0bef94ab9f5d30a72f76c71e8682d32ce198c/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 7]
```diff
@@ -92,6 +106,12 @@ private boolean takeTheDirectPath(@NotNull BlockPos target, int range, boolean o
      */
     public boolean walkToBlock(@NotNull BlockPos target, int range, boolean onMove)
     {
+        if(!target.equals(this.target))
+        {
+            reset();
+            this.target = target;
+        }
+
         final double distanceToPath = worker.getColonyJob() instanceof JobBuilder
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR '?' should be on a new line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AOperator_Wrap)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 8]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
             //If he is on the same Y level as his target and both underground, don't use a proxy. Just don't.
             else if (targetY <= levelDepth)
             {
+                double closestNode = Double.MAX_VALUE;
+                Node lastNode = null;
+                for(Map.Entry<Point2D, Node> node : level.getNodes().entrySet())
+                {
+                    final double distanceToNode = node.getKey().distance(worker.getPosition().getX(), worker.getPosition().getZ());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR Local variable 'distanceToNode' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 9]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
...
+                        currentNode = level.getNode(currentNode.getParent());
+                    }
+
+                    for (int i = nodesToTarget.size() - 1; i >= 0; i--)
+                    {
+                        proxyList.add(nodesToTarget.get(i));
+                    }
+                }
+
+                if(!proxyList.isEmpty())
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1155)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* MAJOR 5 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/d42e69dab4f956bb0cc4a7de515b37eeb36400e9/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/d42e69dab4f956bb0cc4a7de515b37eeb36400e9/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java [hunk 10]
```diff
@@ -254,6 +284,48 @@ else if (targetY <= levelDepth && workerY > levelDepth)
             //If he is on the same Y level as his target and both underground, don't use a proxy. Just don't.
             else if (targetY <= levelDepth)
             {
+                double closestNode = Double.MAX_VALUE;
+                Node lastNode = null;
+                for(final Map.Entry<Point2D, Node> node : level.getNodes().entrySet())
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java]: MINOR Local variable 'node' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 5 issues
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/a9568e046f27d3a244af0f4651f1bbef16c5cc7e/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/a9568e046f27d3a244af0f4651f1bbef16c5cc7e/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 4 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/a9568e046f27d3a244af0f4651f1bbef16c5cc7e/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/a9568e046f27d3a244af0f4651f1bbef16c5cc7e/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 4 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/17b1ae3d7b5b0c93f1cafcd7d42e5ccf3a41ac88/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): This method has 118 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS138)
1. MAJOR [WalkToProxy.java#L216](https://github.com/Minecolonies/minecolonies/blob/17b1ae3d7b5b0c93f1cafcd7d42e5ccf3a41ac88/src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java#L216): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

**Comment by marvin-bitterlich** [comment]
LGTM

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/415/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**